File: javascript/hello-world.js
#!/usr/bin/env node
// bail if we don't have our ENV set:
if (!process.env.JMAP_USERNAME || !process.env.JMAP_TOKEN) {
  console.log("Please set your JMAP_USERNAME and JMAP_TOKEN");
  console.log("JMAP_USERNAME=username JMAP_TOKEN=token node hello-world.js");

  process.exit(1);
}

const hostname = process.env.JMAP_HOSTNAME || "api.fastmail.com";
const username = process.env.JMAP_USERNAME;

const authUrl = `https://${hostname}/.well-known/jmap`;
const headers = {
  "Content-Type": "application/json",
  Authorization: `Bearer ${process.env.JMAP_TOKEN}`,
};

const getSession = async () => {
  const response = await fetch(authUrl, {
    method: "GET",
    headers,
  });
  return response.json();
};

const mailboxQuery = async (apiUrl, accountId) => {
  const response = await fetch(apiUrl, {
    method: "POST",
    headers,
    body: JSON.stringify({
      using: ["urn:ietf:params:jmap:core", "urn:ietf:params:jmap:mail"],
      methodCalls: [
        ["Mailbox/query", { accountId, filter: { name: "Drafts" } }, "a"],
      ],
    }),
  });
  const data = await response.json();

  return await data["methodResponses"][0][1].ids[0];
};

const identityQuery = async (apiUrl, accountId) => {
  const response = await fetch(apiUrl, {
    method: "POST",
    headers,
    body: JSON.stringify({
      using: [
        "urn:ietf:params:jmap:core",
        "urn:ietf:params:jmap:mail",
        "urn:ietf:params:jmap:submission",
      ],
      methodCalls: [["Identity/get", { accountId, ids: null }, "a"]],
    }),
  });
  const data = await response.json();

  return await data["methodResponses"][0][1].list.filter(
    (identity) => identity.email === username
  )[0].id;
};

const draftResponse = async (apiUrl, accountId, draftId, identityId) => {
  const messageBody =
    "Hi! \n\n" +
    "This email may not look like much, but I sent it with JMAP, a protocol \n" +
    "designed to make it easier to manage email, contacts, calendars, and more of \n" +
    "your digital life in general. \n\n" +
    "Pretty cool, right? \n\n" +
    "-- \n" +
    "This email sent from my next-generation email system at Fastmail. \n";

  const draftObject = {
    from: [{ email: username }],
    to: [{ email: username }],
    subject: "Hello, world!",
    keywords: { $draft: true },
    mailboxIds: { [draftId]: true },
    bodyValues: { body: { value: messageBody, charset: "utf-8" } },
    textBody: [{ partId: "body", type: "text/plain" }],
  };

  const response = await fetch(apiUrl, {
    method: "POST",
    headers,
    body: JSON.stringify({
      using: [
        "urn:ietf:params:jmap:core",
        "urn:ietf:params:jmap:mail",
        "urn:ietf:params:jmap:submission",
      ],
      methodCalls: [
        ["Email/set", { accountId, create: { draft: draftObject } }, "a"],
        [
          "EmailSubmission/set",
          {
            accountId,
            onSuccessDestroyEmail: ["#sendIt"],
            create: { sendIt: { emailId: "#draft", identityId } },
          },
          "b",
        ],
      ],
    }),
  });

  const data = await response.json();
  console.log(JSON.stringify(data, null, 2));
};

const run = async () => {
  const session = await getSession();
  const apiUrl = session.apiUrl;
  const accountId = session.primaryAccounts["urn:ietf:params:jmap:mail"];
  const draftId = await mailboxQuery(apiUrl, accountId);
  const identityId = await identityQuery(apiUrl, accountId);
  draftResponse(apiUrl, accountId, draftId, identityId);
};

run();

File: lua/hello_world.lua
#!/usr/bin/env lua

local tiny_jmap = require("tiny_jmap")

-- Set up our client from the environment and set our account ID
local client = tiny_jmap.new({
  hostname = os.getenv("JMAP_HOSTNAME") or "jmap.fastmail.com",
  username = os.getenv("JMAP_USERNAME"),
  password = os.getenv("JMAP_PASSWORD"),
})
local account_id = client:get_account_id()

-- Here, we're going to find our drafts mailbox, by calling Mailbox/query
local query_res = client:make_jmap_call(
  {
    using = {
      "urn:ietf:params:jmap:core",
      "urn:ietf:params:jmap:mail",
    },
    methodCalls = {
      {
        "Mailbox/query",
        {
          accountId = account_id,
          filter = {
            name = "Drafts",
          },
        },
        "a",
      },
    },
  }
)

-- Pull out the id from the list response, and make sure we got it
local draft_mailbox_id = query_res.methodResponses[1][2].ids[1]
assert(draft_mailbox_id)

-- Great! Now we're going to set up the data for the email we're going to send.
local body = [[
Hi!

This email may not look like much, but I sent it with JMAP, a new protocol
designed to make it easier to manage email, contacts, calendars, and more of
your digital life in general.

Pretty cool, right?

-- 
This email sent from my next-generation email system at Fastmail.
]]

local draft = {
  from = {{ email = client.username }},
  to = {{ email = client.username }},
  subject = "Hello, world!",
  keywords = {},
  mailboxIds = {},
  bodyValues = { body = { value = body, charset = "utf-8" } },
  textBody = {{ partId = "body", type = "text/plain" }},
}
draft.keywords["$draft"] = true
draft.mailboxIds[draft_mailbox_id] = true

-- Here, we make two calls in a single request. The first is an Email/set, to
-- set our draft in our drafts folder, and the second is an
-- EmailSubmission/set, to actually send the mail to ourselves. This requires
-- an additional capability for submission.
local create_res = client:make_jmap_call(
  {
    using = {
      "urn:ietf:params:jmap:core",
      "urn:ietf:params:jmap:mail",
      "urn:ietf:params:jmap:submission",
    },
    methodCalls = {
      {
        "Email/set",
        {
          accountId = account_id,
          create = {
            draft = draft,
          },
        },
        "a",
      },
      {
        "EmailSubmission/set",
        {
          accountId = account_id,
          onSuccessDestroyEmail = {"#sendIt"},
          create = { sendIt = { emailId = "#draft" }},
        },
        "b",
      },
    },
  }
)

local serpent = require("serpent")
print(serpent.block(create_res))
